### PR TITLE
Extend logging for dropped warning headers.

### LINKED ISSUE #90527: Identify client for whom warning headers were dropped
Today if a client make a request that generates too many warning headers, Elasticsearch will log a warning message of the form `Dropping a warning header, as their total size reached the maximum allowed ...`. Elasticsearch typically adds warning headers because the client is doing something wrong, but the message doesn't identify the affected client, which makes it hard for users to investigate and address the problem.

Could we add some identifying information (address & port, user agent, ...) about the affected client to messages like these? Presumably we'd need to capture these things on the coordinating node and plumb them around the system via the thread context.

### PR DESCRIPTION
**Pull request by mosche**
This PR adds selected headers (xOpaqueId, product origin) when logging to help better understand the source of requests triggering such warnings.

Closes #90527

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java [hunk 1]
```diff
@@ -798,6 +799,24 @@ private ThreadContextStruct putResponseHeaders(Map<String, Set<String>> headers)
             return new ThreadContextStruct(requestHeaders, newResponseHeaders, transientHeaders, isSystemContext);
         }
 
+        private void logWarningHeaderThresholdExceeded(long threshold, Setting<?> thresholdSetting) {
+            // If available, log some selected headers to help identifying the source of the request.
+            // Note: Only Task.HEADERS_TO_COPY are guaranteed to be preserved at this point.
```
Comment (mosche) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:804]: TBD: From a user's perspective it would be convenient to also include `User-Agent` in here. Though, currently, that isn't part of the headers that are copied over. I'm on the fence if to add it or not.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java [hunk 2]
```diff
@@ -798,6 +799,24 @@ private ThreadContextStruct putResponseHeaders(Map<String, Set<String>> headers)
...
+            Map<String, String> selectedHeaders = new HashMap<>(requestHeaders);
+            selectedHeaders.keySet().retainAll(List.of(Task.X_OPAQUE_ID_HTTP_HEADER, Task.X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER));
+
+            String messagePattern = HttpTransportSettings.SETTING_HTTP_MAX_WARNING_HEADER_SIZE.equals(thresholdSetting)
+                ? "Dropping a warning header{}, as their total size reached the maximum allowed of [{}] bytes set in [{}]!"
+                : "Dropping a warning header{}, as their total count reached the maximum allowed of [{}] set in [{}]!";
+
+            logger.warn(
+                messagePattern,
+                selectedHeaders.isEmpty() ? "" : Strings.format(" for request [%s]", selectedHeaders),
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:814]: Could we mention here that the request had no `X-Opaque-Id` header, to guide users towards setting this header for further investigation without needing to open a support case. Ideally also using `ReferenceDocs` to add a link to [the relevant docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/api-conventions.html#x-opaque-id) too.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:804]: I'd be inclined not to. We don't use this to identify requests elsewhere.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java (hunk 2 above)
Comment (mosche) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:824]: Great suggestion, thanks for the pointer @DaveCTurner

**Review by DaveCTurner — APPROVED** [review]
_kinda_ LGTM although I must admit I find the nest of `String.format()` calls a little hard to follow. Are there neater ways to achieve this? Maybe a `StringBuilder`?

**Review by DaveCTurner — APPROVED** [review]
LGTM thanks for the simplification
